pkg/commands/addmetadata.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package commands
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/kubernetes-sigs/kustomize/pkg/constants"
+	"github.com/kubernetes-sigs/kustomize/pkg/fs"
+	"github.com/spf13/cobra"
+)
+
+// KindOfAdd is the kind of metadata being added: label or annotation
+type KindOfAdd int
+
+const (
+	annotation KindOfAdd = iota
+	label
+)
+
+func (k KindOfAdd) String() string {
+	kinds := [...]string{
+		"annotation",
+		"label",
+	}
+	if k < 0 || k > 1 {
+		return "Unknown metadatakind"
+	}
+	return kinds[k]
+}
+
+type addMetadataOptions struct {
+	metadata map[string]string
+}
+
+// newCmdAddAnnotation adds one or more commonAnnotations to the kustomization file.
+func newCmdAddAnnotation(fsys fs.FileSystem) *cobra.Command {
+	var o addMetadataOptions
+
+	cmd := &cobra.Command{
+		Use:   "annotation",
+		Short: "Adds one or more commonAnnotations to the kustomization.yaml in current directory",
+		Example: `
+		add annotation {annotationKey1:annotationValue1},{annotationKey2:annotationValue2}`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := o.Validate(args, annotation)
+			if err != nil {
+				return err
+			}
+			return o.RunAddAnnotation(fsys, annotation)
+		},
+	}
+	return cmd
+}
+
+// newCmdAddLabel adds one or more commonLabels to the kustomization file.
+func newCmdAddLabel(fsys fs.FileSystem) *cobra.Command {
+	var o addMetadataOptions
+
+	cmd := &cobra.Command{
+		Use:   "label",
+		Short: "Adds one or more commonLabels to the kustomization.yaml in current directory",
+		Example: `
+		add label {labelKey1:labelValue1},{labelKey2:labelValue2}`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := o.Validate(args, label)
+			if err != nil {
+				return err
+			}
+			return o.RunAddLabel(fsys, label)
+		},
+	}
+	return cmd
+}
+
+// Validate validates addLabel and addAnnotation commands.
+func (o *addMetadataOptions) Validate(args []string, k KindOfAdd) error {
+	o.metadata = make(map[string]string)
+	if len(args) < 1 {
+		return fmt.Errorf("must specify %s", k)
+	}
+	if len(args) > 1 {
+		return fmt.Errorf("%ss must be comma-separated, with no spaces. See help text for example", k)
+	}
+	inputs := strings.Split(args[0], ",")
+	for _, input := range inputs {
+		ok, err := regexp.MatchString(`\A([a-zA-Z0-9_.-]+):([a-zA-Z0-9_.-]+)\z`, input)
```

Comment:
would it make sense to just re-use the validation functions from k8s.io/apimachinery/pkg/util/validation like `IsQualifiedName`, `IsValidLabelValue`, etc?

Like we did [here](https://github.com/kubernetes-sigs/kustomize/blob/master/pkg/commands/setnamespace.go#L64)?

Reply:
sigh.  afaik, apimachinery intentionally has no notion of semver, because it's internal code.  If it was in client-go it would be fair game.

As far as client validation goes, the choices are 1) do none (let the server 400 on the entire manifest - the least code solution but not user friendly), 2) do weak form of 1 that disallows egregious errors but leaves details to the server (this shouldn't forbid things that the server would allow, and the current code is too strict for annotation values), 3) try to use the same code as the server (read that code and dip into the libraries).

Regardless, I'm in favor of refactoring this PR a bit to isolate the validation check to some shared utility function (shared between this and other kustomize validations - e.g. namespace), merge this PR, then iterate on _that_ set of validation functions.  

I supposed in the end it will just depend on apimachinery :)

Reply:
@guineveresaenger , nice job.  I'd like to merge this ASAP, and at this point just want to isolate, rather than improve or debate about, the validation code.  

Please make a new package pkg/validate/validate.go, and slap some public functions in there like IsValidLabel, IsValidAnnotation, and call them from here, and just move the regexp as is over there.  Would rather put the code in a new package than have it live in commands, because we might want to use it elsewhere (e.g. from loader, and don't want the loader to depend on commands).

Then we can get this PR in.

Later, please file an issue to revisit the validation package and make it depend on (or not) apimachinery (i.e. address the issues discussed here).

Reply:
@monopole - thanks, I appreciate the discussion and am more than happy to help with the next 
steps.